site/src/contexts/ProxyContext.tsx

```diff
@@ -45,15 +80,26 @@ export const ProxyContext = createContext<ProxyContextValue | undefined>(
  */
 export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
   // Try to load the preferred proxy from local storage.
-  let savedProxy = loadPreferredProxy()
+  const [savedProxy, setUserProxy] = useState<Region | undefined>(
+    loadUserSelectedProxy(),
+  )
+
+  // As the proxies are being loaded, default to using the saved proxy.
+  // If the saved proxy is not valid when the async fetch happens, the
+  // selectedProxy will be updated accordingly.
+  let defaultPreferredProxy: PreferredProxy = {
+    selectedProxy: savedProxy,
+    preferredPathAppURL: savedProxy?.path_app_url.replace(/\/$/, "") || "",
+    preferredWildcardHostname: savedProxy?.wildcard_hostname || "",
+  }
```

Comment:
Would it make sense to just store this whole object in local storage? We could update it easily when we need to with the spread operator, e.g.
````
saveUserSelectedProxy({
      ...proxyObj,
      selectedProxy: newValue
})
````

Reply:
I used to store the entire `PreferredProxy`, but the issue is that `PreferredProxy` is only computed from `getPreferredProxy`. That function takes in a `Region`, and returns the `PreferredProxy`

The returned `PreferredProxy` **does not necessarily match the input region**. So there is no helper function to turn a `Region` -> `PreferredProxy`.

So I just store the region, and let the `getPreferredProxy` compute the extra fields.

Now unfortunately, before all the async stuff happens, I just decide to trust the user's local storage state and need to set the defaults. Maybe the solution is make a function to turn `Region` -> `PreferredProxy` and just call that from both places... I'll do that.